### ISSUE TITLE
Backpropagation algorithm implementation

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#355]
OK, so I'm going through the back-prop implementation, which is implemented in MultiLayerNetwork.doBackward() and BaseLayer.backwardGradient().
There are a number of significant issues here.

First I'm going to define notation and the back-prop algorithm for the sake of discussion/debugging/my own sanity:

Denote:
![](http://mathurl.com/q8nt7yw.png) as weight from unit i (prev. layer) to unit j (output layer).
![](http://mathurl.com/nr9pktp.png) is input to unit j
![](http://mathurl.com/osglszn.png) is target/label
![eqn](http://mathurl.com/oopsolc.png) is output
L loss function
L (superscript): layer number.
![eqn](http://mathurl.com/ojcy77o.png) is element-wise nonlinearity/activation function
![](http://mathurl.com/ooyn3vk.png) is the activation of the jth unit in layer L. Note that ![](http://mathurl.com/pl79ogo.png)

For the output layer, define the error/delta:
![](http://mathurl.com/ngokjpv.png)
(See https://github.com/deeplearning4j/deeplearning4j/issues/349)

For lower layers, we have (with ![](http://mathurl.com/nq6x75r.png) being element-wise multiplication):
![](http://mathurl.com/ovapmc2.png)

(References for above equations:)
(1) http://neuralnetworksanddeeplearning.com/chap2.html (eq. 29, BP1, BP2)
(2) Bishop (1995) Neural Networks for Pattern Recognition (pg144/eq4.36)

-----
-----
MultiLayerNetwork.doBackWard(): Javadoc: "Calculate the gradient relative to the error in the next layer"
Given the context and how the results are used, I'm going to assume this means calculate ![](http://mathurl.com/qx7u6px.png) given ![](http://mathurl.com/on7zv3k.png)

The error/delta/ix for the output layer
```java
INDArray ixInitial = labels.sub(outputActivate)
                    .subi(Nd4j.getExecutioner()
                            .execAndReturn(Nd4j.getOpFactory()
                                    .createTransform(getOutputLayer().conf().getActivationFunction(),outputActivate).derivative()));
```
should simply be:
```java
INDArray ixInitial = outputActivate.sub(labels);
```

Furthermore,
```for(int j = getnLayers() - 1; j >= 0; j--) { ```
becomes
```for(int j = getnLayers() - 2; j >= 0; j--) {```
as we already have the error/delta for the output layer (i.e., the above correction for ixInitial).

From here, it gets a little messy.
As best I can tell, BaseLayer.backwardGradient(activations.get(j),ix) implements the following (with the correct implementation after the notEq symbol):
![](http://mathurl.com/ot6o77t.png)
where
![](http://mathurl.com/p9flynx.png)
![](http://mathurl.com/ok5ajb5.png)
![](http://mathurl.com/q7ho52d.png)

-----
Then, for converting the errors into weight/bias updates in BaseLayer.calcGradient(Gradient layerError, INDArray activation):
![](http://mathurl.com/ndsdfmc.png)
![](http://mathurl.com/q3awqfe.png)

(Ref: http://neuralnetworksanddeeplearning.com/chap2.html, BP4, BP3)
dL/dW looks correct here, but I don't see how dL/db is equivalent to the above equation.
```java
INDArray weightError = weightErrorSignal.transpose().mmul(activation).divi(weightErrorSignal.rows()).transpose();
INDArray biasGradient = weightError.sum(0).divi(weightErrorSignal.rows());
```

**Comment by agibsonccc** [comment]
Good find. A lot changed during the refactoring (trying to generalize back prop on a layer to work with any layer)
It was good in the original BaseMultiLayerNetwork. Let me take a bit of a closer look documenting this. I'll define some baseline terms and get a proof up (or at least double check what's there)

**Comment by agibsonccc** [comment]
So one thing of note here. The book interprets examples as column vectors rather than row vectors like we do.

**Comment by AlexDBlack** [comment]
"The reason for this is due to the fact that I want to separate backpropagation from the coefficients."
Is the motivation here to allow for other architectures that are not simply L->L+1 in terms of connections: i.e., RNNs, networks with skip connections and the like?

"Another thing is the j. The idea is rather than the + 1. Everything that's "current" is actually the coefficient for the next layer."
Still trying to get my head around this.
So suppose I have two layers, layer[1] is OutputLayer, layer[0] is a hidden layer (say, RBM layer in practice).
Just before the for loop, we have errors[1] = iX = ixInitial = (out - labels), which seems correct to me. Then there is only one back-prop step required, to calculate errors[0], right?
As implemented, there are two iterations for two layers instead of 1 (i.e., j=1, j=0), and on the first iteration of the loop (j=1), errors[1] this is overwritten? Though perhaps I've misunderstood something here.

Anyway, to clarify: is the intent is that layer[1] is used to calculate errors[0], right? As opposed to layers[0] used to calculate errors[0]?

Assuming the former, I think (but I'm not sure) that this leads to another issue: my thinking is that with two layers we have two sigmaPrimes, one for the output layer, and one for the hidden layer. Denote them sigmaPrime^1 and sigmaPrime^0.
We are passing z^0 as an argument to BaseLayer.backwardGradient(). Doesn't the current implementation with 2 layers do sigmaPrime^1(z^0) instead of sigmaPrime^0(z^0)? i.e., it's using the output layer activation function derivative, not the previous/hidden layer activation function derivative?

**Comment by agibsonccc** [comment]
So I made some of the changes. I def mis interpreted z.
https://github.com/deeplearning4j/deeplearning4j/blob/backprop/deeplearning4j-core/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L1016

Re: indexing yes. I don't count the output layer. So the idea with backward gradient is to update the weight matrices (not counting the final output in the indices)

The intent is to keep the calculation subcontained since calculating the error is:

error * coefficients

I want to be able to pass in the error and multiply the coefficients (in this case a gradient which could be multiple parameters/ceffocients) in to the layer.

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/deeplearning4j/pull/356

**Comment by agibsonccc** [comment]
@AlexDBlack  https://github.com/deeplearning4j/deeplearning4j/commit/cc253590d91cc29bc5d3a4b188f043d1db3c4313
